scanner/src/main/kotlin/experimental/WorkingTreeCache.kt

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2021 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.experimental
+
+import java.io.IOException
+
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+import org.ossreviewtoolkit.downloader.VersionControlSystem
+import org.ossreviewtoolkit.downloader.WorkingTree
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.utils.common.safeDeleteRecursively
+import org.ossreviewtoolkit.utils.core.createOrtTempDir
+
+/**
+ * A cache for [VCS][VersionControlSystem] [WorkingTree]s that manages one working tree for each triple of
+ * [VcsInfo.type], [VcsInfo.url], and [VcsInfo.path].
+ */
+interface WorkingTreeCache {
+    /**
+     * Apply the [block] to the working tree defined by [vcsInfo]. The [VcsInfo.revision] is ignored during
+     * initialization of the working tree, it is the caller's responsibility to update the working tree to the correct
+     * revision. It is ensured that only one [block] at a time can access a single working tree, separate working trees
+     * can be accessed in parallel.
+     */
+    suspend fun <T> use(vcsInfo: VcsInfo, block: (VersionControlSystem, WorkingTree) -> T): T
+
+    /**
+     * Clear all cached working trees from the file system. The function waits for all currently running operations on
+     * the working trees to finish before deleting them.
+     */
+    suspend fun clear()
+}
+
+class DefaultWorkingTreeCache : WorkingTreeCache {
+    private val mutex = Mutex()
+    private val workingTreeMutexes = mutableMapOf<String, Mutex>()
+    private val workingTrees = mutableMapOf<String, WorkingTree>()
+
+    override suspend fun <T> use(vcsInfo: VcsInfo, block: (VersionControlSystem, WorkingTree) -> T): T =
+        getWorkingTreeMutex(vcsInfo).withLock {
+            val vcs = getVcs(vcsInfo)
+            val workingTree = getWorkingTree(vcsInfo)
+
+            block(vcs, workingTree)
+        }
+
+    private fun getKey(vcsInfo: VcsInfo) = "${vcsInfo.type}|${vcsInfo.url}|${vcsInfo.path}"
```

Comment:
Why is the path part of the key?

Reply:
Two reasons:
* For git-repo the path represents the manifest and I don't want to reuse the same working tree for different manifests.
* For git the sparse checkout is configured durint `initWorkingTree()` so it is currently not possible to reuse the working tree for different paths.

Reply:
In the context of previous refactorings moving towards "scan by repository" it recall that it often times made sense to move away from sparse checkouts. So, would this make sense also here? e.g. drop the path from the vcs info so that no sparse checkout is made?

Reply:
I initially did this, but git-repo made it impossible because we need the path there to know which manifest to use. I would like to change this so that for git-repo the manifest path is given as part of the URL (like `http://path/to/repo?manifest.xml`), then we could drop the path here.

Reply:
>  for git-repo the manifest path is given as part of the URL (like http://path/to/repo?manifest.xml), then we could drop the path here.

nice idea!